---
title: "AnaCoDa_Fall2024_Garcia"
author: "Sarah Garcia"
date: "2024-09-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Objective:

-   Determine if the mutation rate is different between the forward and reverse strands of bacteria.
-   Using E.coli K-12 genome for testing (obtained via Ensembl Bacteria)

### Using AnaCoDa

Load needed libraries

```{r}
library(tidyverse)
library(AnaCoDa)
library(seqinr)
library(bayesplot)
```

First step is to load in the E.coli K-12 strain .fasta file (CDS).

```{r}

#Define .fasta file in working directory
dir <- "Escherichia_coli_str_k_12_substr_w3110_gca_000010245.ASM1024v1.cds.all.fa"

#Define function
chomp<-function(dir){  
  
  genome <- read_delim(dir, delim = "\n", col_names = FALSE, col_types = "c")
  genome.list <- genome %>% na.omit() %>% data.frame()
  
  i=0                                             #Initialize empty objects for faster processing time
  flag=0
  y<-data.frame()
  gene.length<-list()
  sequence.all<-list()
  alist<-list()
  #species.name<-species.list                      #this will be user supplied, must match how it looks in FASTA
  b<-''
  c<-''
  
  cat(paste("Now importing:",dir, "\n"))        #basic progress checker
  for(i in genome.list[ ,1]){                 #for each line of inputted genomeFASTA
    if(str_detect(i,"^>")){                     #if row starts with >
      if(flag==1){                              #flag check to concatenate sequence lines
        sequence<-str_c(alist,collapse = "")    #collapse
        aasize<-nchar(sequence)                 #get amino acid size
        gene.length<-rbind.data.frame(gene.length,aasize)  #store size
        sequence.all<-rbind.data.frame(sequence.all,sequence)  #store sequence
        alist<-list()                           #empty temp variable
        flag=0                                  #reset flag
      }
      flag=0                                    #reset
      pattern.chr<-paste0("Chromosome:\\S+")#make species name pattern
      pattern.gn<-paste0("description:.*")             #make gene name pattern
      pattern.type<-paste0("gene_biotype:\\S+")            #make >tr or >sp pattern 
      a<-str_extract(i,pattern = pattern.chr) #find and store species name (user supplies species name)
      b<-str_extract(i,pattern = pattern.gn)    #find and store gene name
      c<-str_extract(i,pattern = pattern.type)  #find and store pattern type (>tr or >sp, should be only two options)
      abci<-c(a,b,c,i)                             #make above identifiers into easy to bind object
      y<-rbind(y,abci)                           #rbind to fill rows of new df
    }
    else{
      flag=1                                    #Raise the flag!
      b<-str_trim(i,side = c("right"))          #Cut off new line character
      alist[[i]]<-b                             #Storing all sequence lines of current protein
    }
  }
  if(flag==1){                                  #this catches the last protein's sequence VVV
    sequence<-str_c(alist,collapse = "")        #
    aasize<-nchar(sequence)                     #
    gene.length<-rbind.data.frame(gene.length,aasize) #
    sequence.all<-rbind.data.frame(sequence.all,sequence) #
    alist<-list()                               #
    flag=0                                      #^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  }
  y[,5]<-gene.length                             #add sizes to df
  y[,6]<-sequence.all                             #add sequences to df
  colnames(y)<-c("Location","Description","Type","Header","Length","Sequence") #add colnames to object 'y'
  y$Description <- str_replace_all(y$Description, "description:", "")  #remove OS= from species column
  y$Location <- str_replace_all(y$Location, "Chromosome:", "")  #remove GN= from gene name column
  y$Type <- str_replace_all(y$Type, "gene_biotype:", "")
  y$Header <- str_replace_all(y$Header, ">", "")
  
  #Count the number of 'genes' that don't have descriptions
  perc <- (sum(is.na(y$Description)))/nrow(y)*100
  cat("Percentage of individual sequences w/o description:", perc)
  
  genome <<- y
}   #Load 'chomp' function to import .fasta files

#Run function on .fasta file to import
chomp(dir)

```

Next, we will partition the genome by sorting assigning each sequence F (forward) or R (reverse) strand. This depends on using Ensembl Bacteria's assignment (1 or -1).

Source: <https://bacteria.ensembl.org/Escherichia_coli_str_k_12_substr_w3110_gca_000010245/Info/Index>

```{r}

#Take generated genome file and separate into groups based on chromosome strand assignment (1 or -1)
reverse <- genome %>% filter(str_detect(Location, ":-1$")) %>% mutate(pos = "R", assignment = 2)
forward <- genome %>% filter(str_detect(Location, ":1$")) %>% mutate(pos = "F", assignment = 1)
genome.pos <- rbind(reverse, forward)   #This changed the order, which messed up the gene.assignments!
genome.pos <- genome %>% left_join(genome.pos) %>% mutate(id = str_extract(Header, "^[^\\s]+"))

#Determine gene distribution
cat("Genes on forward strand:", nrow(forward), "-->", (nrow(forward)/nrow(genome))*100, "%", 
    "\nGenes on reverse strand:", nrow(reverse), "-->", (nrow(reverse)/nrow(genome))*100, "%")

```

We see that the split of coding sequences is roughly equal across the forward and reverse strands. Now we will begin initializing objects for the MCMC.

```{r warning = FALSE}

genomes <- initializeGenomeObject(
  file = "Escherichia_coli_str_k_12_substr_w3110_gca_000010245.ASM1024v1.cds.all.fa"
  )

parameter <- initializeParameterObject(
  genome = genomes,    #set genome file
  sphi = c(0.5, 2),    #provide phi values
  num.mixtures = 2,    #define number of groups
  gene.assignment = genome.pos$assignment, #define gene assignments
  model = "ROC"        #define model
  )

parameter$fixDEta() #fixing our selection parameter (i.e. assuming neutral selection for strandss)

model <- initializeModelObject(parameter = parameter, model = "ROC")

mcmc <- initializeMCMCObject(samples = 5000, thinning = 10, adaptive.width=50)

```

Run the model [Current run time \~1.5hrs]

```{r}

#runMCMC(mcmc = mcmc, genome = genomes, model = model)

```

Make sure to save your files immediately after the run.

```{r}

#writeParameterObject(parameter = parameter, file = "parameter.file.name.Rda")
#writeMCMCObject(mcmc = mcmc, file = "MCMC.file.name.Rda")

```

Use the following code to load in a previous run.

```{r}

#Use this code to load it back in
#parameter <- loadParameterObject(files = "parameter_out_Ecoli_split_09112024.Rda")
mcmc <- loadMCMCObject(file = "mcmc_out_Ecoli_split_09162024.Rda")

### This code was used to import parameter file an alternative way
  #i <- 1
  files <- "parameter_out_Ecoli_split_09162024.Rda"
  
  #load("parameter_out_Ecoli_split_09112024.Rda")
  #This function is currently not working - FIXED
  loadROCParameterObject <- function(parameter, files){
    
    setBaseInfo <- function(parameter, files){
      for (i in 1:length(files)) {
        tempEnv <- new.env();
        load(file = files[i], envir = tempEnv)
        if (i == 1) {
          categories <- tempEnv$paramBase$categories
          categories.matrix <- do.call("rbind", tempEnv$paramBase$categories)
          numMixtures <- tempEnv$paramBase$numMix
          numMutationCategories <- tempEnv$paramBase$numMut
          numSelectionCategories <- tempEnv$paramBase$numSel
          mixtureAssignment <- tempEnv$paramBase$curMixAssignment
          lastIteration <- tempEnv$paramBase$lastIteration
          max <- tempEnv$paramBase$lastIteration + 1
          grouplist <- tempEnv$paramBase$grouplist
          
          stdDevSynthesisRateTraces <- vector("list", length = numSelectionCategories)
          for (j in 1:numSelectionCategories) {
            stdDevSynthesisRateTraces[[j]] <- tempEnv$paramBase$stdDevSynthesisRateTraces[[j]][1:max]
          }
          stdDevSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$stdDevSynthesisRateAcceptRatTrace
          synthesisRateTrace <- vector("list", length = numSelectionCategories)
          for (j in 1:numSelectionCategories) {
            for (k in 1:length(tempEnv$paramBase$synthRateTrace[[j]])){
              synthesisRateTrace[[j]][[k]] <- tempEnv$paramBase$synthRateTrace[[j]][[k]][1:max]
            }
          }
          synthesisRateAcceptanceRateTrace <- tempEnv$paramBase$synthAcceptRatTrace
          mixtureAssignmentTrace <- vector("list", length = length(tempEnv$paramBase$mixAssignTrace))
          for (j in 1:length(tempEnv$paramBase$mixAssignTrace)){
            mixtureAssignmentTrace[[j]] <- tempEnv$paramBase$mixAssignTrace[[j]][1:max]
          }
          mixtureProbabilitiesTrace <- c()
          for (j in 1:numMixtures) {
            mixtureProbabilitiesTrace[[j]] <- tempEnv$paramBase$mixProbTrace[[j]][1:max]
          }
          codonSpecificAcceptanceRateTrace <- tempEnv$paramBase$codonSpecificAcceptRatTrace
          
          ### ERROR HERE ###
          withPhi <- tempEnv$paramBase$withPhi
          if (withPhi){
            phiGroups <- length(tempEnv$paramBase$synthesisOffsetTrace) #add $paramBase
            synthesisOffsetTrace <- c()
            for (j in 1:phiGroups) {
              synthesisOffsetTrace[[j]] <- tempEnv$paramBase$synthesisOffsetTrace[[j]][1:max]
            }
            
            
            synthesisOffsetAcceptanceRateTrace <- tempEnv$paramBase$synthesisOffsetAcceptRatTrace
            
            
            observedSynthesisNoiseTrace <- c()
            for (j in 1:phiGroups) {
              observedSynthesisNoiseTrace[[j]] <- tempEnv$paramBase$observedSynthesisNoiseTrace[[j]][1:max]
            }
            #need number of phi groups, not the number of mixtures apparently.
          }else {
            synthesisOffsetTrace <- c()
            synthesisOffsetAcceptanceRateTrace <- c()
            observedSynthesisNoiseTrace <- c()
          }
        } else {
          if (sum(categories.matrix != do.call("rbind", tempEnv$paramBase$categories)) != 0){
            stop("categories is not the same between all files")
          }#end of error check
          
          if (numMixtures != tempEnv$paramBase$numMix){
            stop("The number of mixtures is not the same between files")
          }
          
          if (numMutationCategories != tempEnv$paramBase$numMut){
            stop("The number of mutation categories is not the same between files")
          }
          
          if (numSelectionCategories != tempEnv$paramBase$numSel){
            stop("The number of selection categories is not the same between files")
          }
          
          if (length(mixtureAssignment) != length(tempEnv$paramBase$curMixAssignment)){
            stop("The length of the mixture assignment is not the same between files. 
               Make sure the same genome is used on each run.")
          }
          
          if(length(grouplist) != length(tempEnv$paramBase$grouplist)){
            stop("Number of Amino Acids/Codons is not the same between files.")	
          }
          if (withPhi != tempEnv$paramBase$withPhi){
            stop("Runs do not match in concern in with.phi")
          }
          
          curSynthesisOffsetTrace <- tempEnv$paramBase$synthesisOffsetTrace
          curSynthesisOffsetAcceptanceRateTrace <- tempEnv$paramBase$synthesisOffsetAcceptRatTrace
          curObservedSynthesisNoiseTrace <- tempEnv$paramBase$observedSynthesisNoiseTrace
          
          if (withPhi){
            combineTwoDimensionalTrace(synthesisOffsetTrace, curSynthesisOffsetTrace, max)
            size <- length(curSynthesisOffsetAcceptanceRateTrace)
            combineTwoDimensionalTrace(synthesisOffsetAcceptanceRateTrace, curSynthesisOffsetAcceptanceRateTrace, size)
            combineTwoDimensionalTrace(observedSynthesisNoiseTrace, curObservedSynthesisNoiseTrace, max)
          }
          
          
          curStdDevSynthesisRateTraces <- tempEnv$paramBase$stdDevSynthesisRateTraces
          curStdDevSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$stdDevSynthesisRateAcceptRatTrace
          curSynthesisRateTrace <- tempEnv$paramBase$synthRateTrace
          curSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$synthAcceptRatTrace
          curMixtureAssignmentTrace <- tempEnv$paramBase$mixAssignTrace
          curMixtureProbabilitiesTrace <- tempEnv$paramBase$mixProbTrace
          curCodonSpecificAcceptanceRateTrace <- tempEnv$paramBase$codonSpecificAcceptRatTrace
          
          lastIteration <- lastIteration + tempEnv$paramBase$lastIteration
          
          
          #assuming all checks have passed, time to concatenate traces
          max <- tempEnv$paramBase$lastIteration + 1
          combineTwoDimensionalTrace(stdDevSynthesisRateTraces, curStdDevSynthesisRateTraces, max)
          
          size <- length(curStdDevSynthesisRateAcceptanceRateTrace)
          stdDevSynthesisRateAcceptanceRateTrace <- c(stdDevSynthesisRateAcceptanceRateTrace, 
                                                      curStdDevSynthesisRateAcceptanceRateTrace[2:size])
          
          
          combineThreeDimensionalTrace(synthesisRateTrace, curSynthesisRateTrace, max)
          size <- length(curSynthesisRateAcceptanceRateTrace)
          combineThreeDimensionalTrace(synthesisRateAcceptanceRateTrace, curSynthesisRateAcceptanceRateTrace, size)
          
          combineTwoDimensionalTrace(mixtureAssignmentTrace, curMixtureAssignmentTrace, max)
          combineTwoDimensionalTrace(mixtureProbabilitiesTrace, curMixtureProbabilitiesTrace, max)
          size <- length(curCodonSpecificAcceptanceRateTrace)
          combineTwoDimensionalTrace(codonSpecificAcceptanceRateTrace, curCodonSpecificAcceptanceRateTrace, size)
        }
      }
      
      parameter$setCategories(categories)
      parameter$setCategoriesForTrace()  
      parameter$numMixtures <- numMixtures
      parameter$numMutationCategories <- numMutationCategories
      parameter$numSelectionCategories <- numSelectionCategories
      parameter$setMixtureAssignment(tempEnv$paramBase$curMixAssignment) #want the last in the file sequence
      parameter$setLastIteration(lastIteration)
      parameter$setGroupList(grouplist)
      
      trace <- parameter$getTraceObject()
      trace$setStdDevSynthesisRateTraces(stdDevSynthesisRateTraces)
      trace$setStdDevSynthesisRateAcceptanceRateTrace(stdDevSynthesisRateAcceptanceRateTrace)
      trace$setSynthesisRateTrace(synthesisRateTrace)
      trace$setSynthesisRateAcceptanceRateTrace(synthesisRateAcceptanceRateTrace)
      trace$setSynthesisOffsetTrace(synthesisOffsetTrace)
      trace$setSynthesisOffsetAcceptanceRateTrace(synthesisOffsetAcceptanceRateTrace)
      trace$setObservedSynthesisNoiseTrace(observedSynthesisNoiseTrace)
      trace$setMixtureAssignmentTrace(mixtureAssignmentTrace)
      trace$setMixtureProbabilitiesTrace(mixtureProbabilitiesTrace)
      trace$setCodonSpecificAcceptanceRateTrace(codonSpecificAcceptanceRateTrace)
      
      parameter$setTraceObject(trace)
      return(parameter)
    } #changed a single line
    
    parameter <- setBaseInfo(parameter, files)
    for (i in 1:length(files)){
      tempEnv <- new.env();
      load(file = files[i], envir = tempEnv)
      
      numMutationCategories <- tempEnv$paramBase$numMut
      numSelectionCategories <- tempEnv$paramBase$numSel
      max <- tempEnv$paramBase$lastIteration + 1
      
      if (i == 1){
        
        
        codonSpecificParameterTraceMut <- vector("list", length=numMutationCategories)
        for (j in 1:numMutationCategories) {
          codonSpecificParameterTraceMut[[j]] <- vector("list", length=length(tempEnv$mutationTrace[[j]]))
          for (k in 1:length(tempEnv$mutationTrace[[j]])){
            codonSpecificParameterTraceMut[[j]][[k]] <- tempEnv$mutationTrace[[j]][[k]][1:max]
          }
        }
        
        codonSpecificParameterTraceSel <- vector("list", length=numSelectionCategories)
        for (j in 1:numSelectionCategories) {
          codonSpecificParameterTraceSel[[j]] <- vector("list", length=length(tempEnv$selectionTrace[[j]]))
          for (k in 1:length(tempEnv$selectionTrace[[j]])){
            codonSpecificParameterTraceSel[[j]][[k]] <- tempEnv$selectionTrace[[j]][[k]][1:max]
          }
        }
      }else{
        
        curCodonSpecificParameterTraceMut <- tempEnv$mutationTrace
        curCodonSpecificParameterTraceSel <- tempEnv$selectionTrace
        combineThreeDimensionalTrace(codonSpecificParameterTraceMut, curCodonSpecificParameterTraceMut, max)
        combineThreeDimensionalTrace(codonSpecificParameterTraceSel, curCodonSpecificParameterTraceSel, max)
      }#end of if-else
    }#end of for loop (files)
    
    trace <- parameter$getTraceObject()
    
    trace$setCodonSpecificParameterTrace(codonSpecificParameterTraceMut, 0)
    trace$setCodonSpecificParameterTrace(codonSpecificParameterTraceSel, 1)
    
    parameter$currentMutationParameter <- tempEnv$currentMutation
    parameter$currentSelectionParameter <- tempEnv$currentSelection
    parameter$proposedMutationParameter <- tempEnv$proposedMutation
    parameter$proposedSelectionParameter <- tempEnv$proposedSelection
    parameter$setTraceObject(trace)
    return(parameter) 
  }
  
  parameter <- loadROCParameterObject(parameter, files)
###

```
