---
title: "AnaCoDa_Fall2024_Garcia"
author: "Sarah Garcia"
date: "2024-09-17"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Objective:

-   Determine if the mutation rate is different between the forward and reverse strands of bacteria.
-   Using E.coli K-12 genome for testing (obtained via Ensembl Bacteria)

### Using AnaCoDa

Load needed libraries

```{r}
library(tidyverse)
library(AnaCoDa)
library(seqinr)
library(bayesplot)

```

Load needed functions - long code chunk, minimize for ease of viewing after loading in functions

```{r, collapse=TRUE}

#Define function
chomp<-function(dir){  
  
  genome <- read_delim(dir, delim = "\n", col_names = FALSE, col_types = "c")
  genome.list <- genome %>% na.omit() %>% data.frame()
  
  i=0                                             #Initialize empty objects for faster processing time
  flag=0
  y<-data.frame()
  gene.length<-list()
  sequence.all<-list()
  alist<-list()
  #species.name<-species.list                      #this will be user supplied, must match how it looks in FASTA
  b<-''
  c<-''
  
  cat(paste("Now importing:",dir, "\n"))        #basic progress checker
  for(i in genome.list[ ,1]){                 #for each line of inputted genomeFASTA
    if(str_detect(i,"^>")){                     #if row starts with >
      if(flag==1){                              #flag check to concatenate sequence lines
        sequence<-str_c(alist,collapse = "")    #collapse
        aasize<-nchar(sequence)                 #get amino acid size
        gene.length<-rbind.data.frame(gene.length,aasize)  #store size
        sequence.all<-rbind.data.frame(sequence.all,sequence)  #store sequence
        alist<-list()                           #empty temp variable
        flag=0                                  #reset flag
      }
      flag=0                                    #reset
      pattern.chr<-paste0("Chromosome:\\S+")#make species name pattern
      pattern.gn<-paste0("description:.*")             #make gene name pattern
      pattern.type<-paste0("gene_biotype:\\S+")            #make >tr or >sp pattern 
      a<-str_extract(i,pattern = pattern.chr) #find and store species name (user supplies species name)
      b<-str_extract(i,pattern = pattern.gn)    #find and store gene name
      c<-str_extract(i,pattern = pattern.type)  #find and store pattern type (>tr or >sp, should be only two options)
      abci<-c(a,b,c,i)                             #make above identifiers into easy to bind object
      y<-rbind(y,abci)                           #rbind to fill rows of new df
    }
    else{
      flag=1                                    #Raise the flag!
      b<-str_trim(i,side = c("right"))          #Cut off new line character
      alist[[i]]<-b                             #Storing all sequence lines of current protein
    }
  }
  if(flag==1){                                  #this catches the last protein's sequence VVV
    sequence<-str_c(alist,collapse = "")        #
    aasize<-nchar(sequence)                     #
    gene.length<-rbind.data.frame(gene.length,aasize) #
    sequence.all<-rbind.data.frame(sequence.all,sequence) #
    alist<-list()                               #
    flag=0                                      #^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  }
  y[,5]<-gene.length                             #add sizes to df
  y[,6]<-sequence.all                             #add sequences to df
  colnames(y)<-c("Location","Description","Type","Header","Length","Sequence") #add colnames to object 'y'
  y$Description <- str_replace_all(y$Description, "description:", "")  #remove OS= from species column
  y$Location <- str_replace_all(y$Location, "Chromosome:", "")  #remove GN= from gene name column
  y$Type <- str_replace_all(y$Type, "gene_biotype:", "")
  y$Header <- str_replace_all(y$Header, ">", "")
  
  #Count the number of 'genes' that don't have descriptions
  perc <- (sum(is.na(y$Description)))/nrow(y)*100
  cat("Percentage of individual sequences w/o description:", perc)
  
  genome <<- y
}   #Load 'chomp' function to import .fasta files

#load("parameter_out_Ecoli_split_09112024.Rda")
loadROCParameterObject <- function(parameter, files){
    
    setBaseInfo <- function(parameter, files){
      for (i in 1:length(files)) {
        tempEnv <- new.env();
        load(file = files[i], envir = tempEnv)
        if (i == 1) {
          categories <- tempEnv$paramBase$categories
          categories.matrix <- do.call("rbind", tempEnv$paramBase$categories)
          numMixtures <- tempEnv$paramBase$numMix
          numMutationCategories <- tempEnv$paramBase$numMut
          numSelectionCategories <- tempEnv$paramBase$numSel
          mixtureAssignment <- tempEnv$paramBase$curMixAssignment
          lastIteration <- tempEnv$paramBase$lastIteration
          max <- tempEnv$paramBase$lastIteration + 1
          grouplist <- tempEnv$paramBase$grouplist
          
          stdDevSynthesisRateTraces <- vector("list", length = numSelectionCategories)
          for (j in 1:numSelectionCategories) {
            stdDevSynthesisRateTraces[[j]] <- tempEnv$paramBase$stdDevSynthesisRateTraces[[j]][1:max]
          }
          stdDevSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$stdDevSynthesisRateAcceptRatTrace
          synthesisRateTrace <- vector("list", length = numSelectionCategories)
          for (j in 1:numSelectionCategories) {
            for (k in 1:length(tempEnv$paramBase$synthRateTrace[[j]])){
              synthesisRateTrace[[j]][[k]] <- tempEnv$paramBase$synthRateTrace[[j]][[k]][1:max]
            }
          }
          synthesisRateAcceptanceRateTrace <- tempEnv$paramBase$synthAcceptRatTrace
          mixtureAssignmentTrace <- vector("list", length = length(tempEnv$paramBase$mixAssignTrace))
          for (j in 1:length(tempEnv$paramBase$mixAssignTrace)){
            mixtureAssignmentTrace[[j]] <- tempEnv$paramBase$mixAssignTrace[[j]][1:max]
          }
          mixtureProbabilitiesTrace <- c()
          for (j in 1:numMixtures) {
            mixtureProbabilitiesTrace[[j]] <- tempEnv$paramBase$mixProbTrace[[j]][1:max]
          }
          codonSpecificAcceptanceRateTrace <- tempEnv$paramBase$codonSpecificAcceptRatTrace
          
          ### ERROR HERE ###
          withPhi <- tempEnv$paramBase$withPhi
          if (withPhi){
            phiGroups <- length(tempEnv$paramBase$synthesisOffsetTrace) #add $paramBase
            synthesisOffsetTrace <- c()
            for (j in 1:phiGroups) {
              synthesisOffsetTrace[[j]] <- tempEnv$paramBase$synthesisOffsetTrace[[j]][1:max]
            }
            
            
            synthesisOffsetAcceptanceRateTrace <- tempEnv$paramBase$synthesisOffsetAcceptRatTrace
            
            
            observedSynthesisNoiseTrace <- c()
            for (j in 1:phiGroups) {
              observedSynthesisNoiseTrace[[j]] <- tempEnv$paramBase$observedSynthesisNoiseTrace[[j]][1:max]
            }
            #need number of phi groups, not the number of mixtures apparently.
          }else {
            synthesisOffsetTrace <- c()
            synthesisOffsetAcceptanceRateTrace <- c()
            observedSynthesisNoiseTrace <- c()
          }
        } else {
          if (sum(categories.matrix != do.call("rbind", tempEnv$paramBase$categories)) != 0){
            stop("categories is not the same between all files")
          }#end of error check
          
          if (numMixtures != tempEnv$paramBase$numMix){
            stop("The number of mixtures is not the same between files")
          }
          
          if (numMutationCategories != tempEnv$paramBase$numMut){
            stop("The number of mutation categories is not the same between files")
          }
          
          if (numSelectionCategories != tempEnv$paramBase$numSel){
            stop("The number of selection categories is not the same between files")
          }
          
          if (length(mixtureAssignment) != length(tempEnv$paramBase$curMixAssignment)){
            stop("The length of the mixture assignment is not the same between files. 
               Make sure the same genome is used on each run.")
          }
          
          if(length(grouplist) != length(tempEnv$paramBase$grouplist)){
            stop("Number of Amino Acids/Codons is not the same between files.")	
          }
          if (withPhi != tempEnv$paramBase$withPhi){
            stop("Runs do not match in concern in with.phi")
          }
          
          curSynthesisOffsetTrace <- tempEnv$paramBase$synthesisOffsetTrace
          curSynthesisOffsetAcceptanceRateTrace <- tempEnv$paramBase$synthesisOffsetAcceptRatTrace
          curObservedSynthesisNoiseTrace <- tempEnv$paramBase$observedSynthesisNoiseTrace
          
          if (withPhi){
            combineTwoDimensionalTrace(synthesisOffsetTrace, curSynthesisOffsetTrace, max)
            size <- length(curSynthesisOffsetAcceptanceRateTrace)
            combineTwoDimensionalTrace(synthesisOffsetAcceptanceRateTrace, curSynthesisOffsetAcceptanceRateTrace, size)
            combineTwoDimensionalTrace(observedSynthesisNoiseTrace, curObservedSynthesisNoiseTrace, max)
          }
          
          
          curStdDevSynthesisRateTraces <- tempEnv$paramBase$stdDevSynthesisRateTraces
          curStdDevSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$stdDevSynthesisRateAcceptRatTrace
          curSynthesisRateTrace <- tempEnv$paramBase$synthRateTrace
          curSynthesisRateAcceptanceRateTrace <- tempEnv$paramBase$synthAcceptRatTrace
          curMixtureAssignmentTrace <- tempEnv$paramBase$mixAssignTrace
          curMixtureProbabilitiesTrace <- tempEnv$paramBase$mixProbTrace
          curCodonSpecificAcceptanceRateTrace <- tempEnv$paramBase$codonSpecificAcceptRatTrace
          
          lastIteration <- lastIteration + tempEnv$paramBase$lastIteration
          
          
          #assuming all checks have passed, time to concatenate traces
          max <- tempEnv$paramBase$lastIteration + 1
          combineTwoDimensionalTrace(stdDevSynthesisRateTraces, curStdDevSynthesisRateTraces, max)
          
          size <- length(curStdDevSynthesisRateAcceptanceRateTrace)
          stdDevSynthesisRateAcceptanceRateTrace <- c(stdDevSynthesisRateAcceptanceRateTrace, 
                                                      curStdDevSynthesisRateAcceptanceRateTrace[2:size])
          
          
          combineThreeDimensionalTrace(synthesisRateTrace, curSynthesisRateTrace, max)
          size <- length(curSynthesisRateAcceptanceRateTrace)
          combineThreeDimensionalTrace(synthesisRateAcceptanceRateTrace, curSynthesisRateAcceptanceRateTrace, size)
          
          combineTwoDimensionalTrace(mixtureAssignmentTrace, curMixtureAssignmentTrace, max)
          combineTwoDimensionalTrace(mixtureProbabilitiesTrace, curMixtureProbabilitiesTrace, max)
          size <- length(curCodonSpecificAcceptanceRateTrace)
          combineTwoDimensionalTrace(codonSpecificAcceptanceRateTrace, curCodonSpecificAcceptanceRateTrace, size)
        }
      }
      
      parameter$setCategories(categories)
      parameter$setCategoriesForTrace()  
      parameter$numMixtures <- numMixtures
      parameter$numMutationCategories <- numMutationCategories
      parameter$numSelectionCategories <- numSelectionCategories
      parameter$setMixtureAssignment(tempEnv$paramBase$curMixAssignment) #want the last in the file sequence
      parameter$setLastIteration(lastIteration)
      parameter$setGroupList(grouplist)
      
      trace <- parameter$getTraceObject()
      trace$setStdDevSynthesisRateTraces(stdDevSynthesisRateTraces)
      trace$setStdDevSynthesisRateAcceptanceRateTrace(stdDevSynthesisRateAcceptanceRateTrace)
      trace$setSynthesisRateTrace(synthesisRateTrace)
      trace$setSynthesisRateAcceptanceRateTrace(synthesisRateAcceptanceRateTrace)
      trace$setSynthesisOffsetTrace(synthesisOffsetTrace)
      trace$setSynthesisOffsetAcceptanceRateTrace(synthesisOffsetAcceptanceRateTrace)
      trace$setObservedSynthesisNoiseTrace(observedSynthesisNoiseTrace)
      trace$setMixtureAssignmentTrace(mixtureAssignmentTrace)
      trace$setMixtureProbabilitiesTrace(mixtureProbabilitiesTrace)
      trace$setCodonSpecificAcceptanceRateTrace(codonSpecificAcceptanceRateTrace)
      
      parameter$setTraceObject(trace)
      return(parameter)
    } #changed a single line
    
    parameter <- setBaseInfo(parameter, files)
    for (i in 1:length(files)){
      tempEnv <- new.env();
      load(file = files[i], envir = tempEnv)
      
      numMutationCategories <- tempEnv$paramBase$numMut
      numSelectionCategories <- tempEnv$paramBase$numSel
      max <- tempEnv$paramBase$lastIteration + 1
      
      if (i == 1){
        
        
        codonSpecificParameterTraceMut <- vector("list", length=numMutationCategories)
        for (j in 1:numMutationCategories) {
          codonSpecificParameterTraceMut[[j]] <- vector("list", length=length(tempEnv$mutationTrace[[j]]))
          for (k in 1:length(tempEnv$mutationTrace[[j]])){
            codonSpecificParameterTraceMut[[j]][[k]] <- tempEnv$mutationTrace[[j]][[k]][1:max]
          }
        }
        
        codonSpecificParameterTraceSel <- vector("list", length=numSelectionCategories)
        for (j in 1:numSelectionCategories) {
          codonSpecificParameterTraceSel[[j]] <- vector("list", length=length(tempEnv$selectionTrace[[j]]))
          for (k in 1:length(tempEnv$selectionTrace[[j]])){
            codonSpecificParameterTraceSel[[j]][[k]] <- tempEnv$selectionTrace[[j]][[k]][1:max]
          }
        }
      }else{
        
        curCodonSpecificParameterTraceMut <- tempEnv$mutationTrace
        curCodonSpecificParameterTraceSel <- tempEnv$selectionTrace
        combineThreeDimensionalTrace(codonSpecificParameterTraceMut, curCodonSpecificParameterTraceMut, max)
        combineThreeDimensionalTrace(codonSpecificParameterTraceSel, curCodonSpecificParameterTraceSel, max)
      }#end of if-else
    }#end of for loop (files)
    
    trace <- parameter$getTraceObject()
    
    trace$setCodonSpecificParameterTrace(codonSpecificParameterTraceMut, 0)
    trace$setCodonSpecificParameterTrace(codonSpecificParameterTraceSel, 1)
    
    parameter$currentMutationParameter <- tempEnv$currentMutation
    parameter$currentSelectionParameter <- tempEnv$currentSelection
    parameter$proposedMutationParameter <- tempEnv$proposedMutation
    parameter$proposedSelectionParameter <- tempEnv$proposedSelection
    parameter$setTraceObject(trace)
    return(parameter) 
  }

```

Now, we load in our CDS genome file. Example uses E.coli K-12 strain .fasta file (CDS).

```{r}

#Define .fasta file in working directory
dir <- "Escherichia_coli_str_k_12_substr_w3110_gca_000010245.ASM1024v1.cds.all.fa"

#Run function on .fasta file to import
chomp(dir)

```

Next, we will partition the genome by sorting assigning each sequence F (forward) or R (reverse) strand. This depends on using Ensembl Bacteria's assignment (1 or -1).

Source: <https://bacteria.ensembl.org/Escherichia_coli_str_k_12_substr_w3110_gca_000010245/Info/Index>

```{r}

#Take generated genome file and separate into groups based on chromosome strand assignment (1 or -1)
reverse <- genome %>% filter(str_detect(Location, ":-1$")) %>% mutate(pos = "R", assignment = 2)
forward <- genome %>% filter(str_detect(Location, ":1$")) %>% mutate(pos = "F", assignment = 1)
genome.pos <- rbind(reverse, forward)   #This changed the order, which messed up the gene.assignments!
genome.pos <- genome %>% left_join(genome.pos) %>% mutate(id = str_extract(Header, "^[^\\s]+"))

#Determine gene distribution
cat("Genes on forward strand:", nrow(forward), "-->", (nrow(forward)/nrow(genome))*100, "%",
    "\nGenes on reverse strand:", nrow(reverse), "-->", (nrow(reverse)/nrow(genome))*100, "%")

```

We see that the split of coding sequences is roughly equal across the forward and reverse strands. Now we will begin initializing objects for the MCMC.

```{r warning = FALSE}

genomes <- initializeGenomeObject(
  file = "Escherichia_coli_str_k_12_substr_w3110_gca_000010245.ASM1024v1.cds.all.fa"
  )

parameter <- initializeParameterObject(
  genome = genomes,    #set genome file
  sphi = c(1, 1),    #provide phi values
  num.mixtures = 2,    #define number of groups
  gene.assignment = genome.pos$assignment, #define gene assignments
  model = "ROC"        #define model
  )

model <- initializeModelObject(parameter = parameter, model = "ROC")

mcmc <- initializeMCMCObject(samples = 5000, thinning = 10, adaptive.width=50)

#Set restart settings
setRestartSettings(mcmc = mcmc, filename = "restart_out_Ecoli_split_sharedSelection_sphi1_09302024", 
                   samples = 2500)


```

Run the model [Current run time \~1.5hrs]

```{r}

#runMCMC(mcmc = mcmc, genome = genomes, model = model)

```

Make sure to save your files immediately after the run.

```{r}

#writeParameterObject(parameter = parameter, file = "parameter.file.name.Rda")
#writeMCMCObject(mcmc = mcmc, file = "MCMC.file.name.Rda")

```

Use the following code to load in a previous run.

```{r}

mcmc <- loadMCMCObject(file = "mcmc_out_Ecoli_split_sharedSelection_fixSphi_09292024.Rda")

files <- "parameter_out_Ecoli_split_sharedSelection_fixSphi_09292024.Rda"
  
parameter <- loadROCParameterObject(parameter, files)

```

Now that we have our model run, let's see if we've reached convergence. The graphs are large, so you might get some errors about dimensions.

```{r, fig.width=10, fig.height=10}

trace <- getTrace(parameter)  #store the trace (i.e. record of all iterations)

#Visualize the mutation parameters
#We are looking for the lines to level out.
plot.diag.1m <- plot(x = trace, what = "Mutation", mixture = 1)
plot.diag.2m <- plot(x = trace, what = "Mutation", mixture = 2)

#We can also visualize the selection parameter
plot.diag.1s <- plot(x = trace, what = "Selection", mixture = 1)
plot.diag.2s <- plot(x = trace, what = "Selection", mixture = 2)

```

The plots above should show each trace becoming more and more level. If not, increase the number of smaples to run during the MCMC. Now we can visualize codon use proportion as it relates to protein synthesis rate phi. We can also visualize how selection and mutation parameters correlated between our mixtures.

```{r, fig.width=10, fig.height=10}

#visualize the results of the model fit
plot(x = model, genome = genomes, samples = 3000, mixture = 1)
plot(x = model, genome = genomes, samples = 3000, mixture = 2)

#This will compare different 'mixtures' i.e. gene sets
plot(parameter, what = "Selection", samples = 3000)
plot(parameter, what = "Mutation", samples = 3000)

```

Now, we want to extract the mutation trace for each individual codon.

```{r}

#Now to grab codon specific parameters
mutationTrace <- trace$getCodonSpecificParameterTrace(0)

#Take object of traces and feed into bayesplot
names(mutationTrace) <- c("mix1", "mix2")

#Get list of codons
names.aa <- aminoAcids()
names.aa <- setdiff(names.aa, c("W", "X", "M")) #remove W, X, and M

#This should be the order conserved in mutationTrace[[i]]
AA.df <- names.aa %>%                 #pipe amino acids
  data.frame(aa = names.aa) %>%       #make a dataframe
  select(-.) %>%                      #remove duplicated column
  mutate(Codon = lapply(aa, AAToCodon)) %>% #make new column with associated codons for each amino acid
  mutate(Codon = map(Codon, ~ .x[-length(.x)])) %>% #remove the reference codon (last alphabetical codon for each AA)
  unnest(cols = c(Codon))                            #unnest column data

#This assumes that the codons are sorted by amino acid, then codon sequence (i.e. A: GCA, GCC, GCG, etc.)
names(mutationTrace[[1]]) <- AA.df$Codon
names(mutationTrace[[2]]) <- AA.df$Codon

#To verify this data matches our plotted trace data, plot a few codons individually to compare
plot(mutationTrace$mix1$AGA, main = "Mixture 1: AGA", ylab = expression(Delta * M), xlab = "Samples")
plot(mutationTrace$mix1$AGC, main = "Mixture 1: AGC", ylab = expression(Delta * M), xlab = "Samples")
plot(mutationTrace$mix1$TTC, main = "Mixture 1: TTC", ylab = expression(Delta * M), xlab = "Samples")

plot(mutationTrace$mix2$AGA, main = "Mixture 2: AGA", ylab = expression(Delta * M), xlab = "Samples")
plot(mutationTrace$mix2$AGC, main = "Mixture 2: AGC", ylab = expression(Delta * M), xlab = "Samples")
plot(mutationTrace$mix2$TTC, main = "Mixture 2: TTC", ylab = expression(Delta * M), xlab = "Samples")

```

Now that we have the posterior distributions of our estimated parameter, let's try to plot it using Bayesplot.

```{r}

#Making trace data frames and discard burn-in
mutTrace.1 <- mutationTrace$mix1 %>% data.frame() %>% slice(-1:-2000)
mutTrace.2 <- mutationTrace$mix2 %>% data.frame() %>% slice(-1:-2000)

#Moving on to Bayesplot
posterior.1 <- as.matrix(mutTrace.1)                     #fit trace into posterior matrix
posterior.2 <- as.matrix(mutTrace.2)

mcmc_areas(posterior.1,                                  #supply the posterior matrix
           pars = c("GCA", "GCC", "GCG"),                #supply the desired traces to plot
           prob = 0.8) +                                 #supply desired interval
  ggtitle("Posterior distributions: Codons of Alanine [Mixture 1]",  #supply title
          "with medians and 80% intervals")# +
  #coord_cartesian(xlim = c(-1.5, 0.5))                   #adjust x-axis

mcmc_areas(posterior.2,                                  #supply the posterior matrix
           pars = c("GCA", "GCC", "GCG"),                #supply the desired traces to plot
           prob = 0.8) +                                 #supply desired interval
  ggtitle("Posterior distributions: Codons of Alanine [Mixture 2]",  #supply title
          "with medians and 80% intervals")# +
  #coord_cartesian(xlim = c(-1.5, 0.5))                   #adjust x-axis

```

We can also directly compare codons across mixtures

```{r}

#Example: show all codons of alanine across the mixtures
alanine <- data.frame(
  GCA.1 = mutTrace.1$GCA,
  GCA.2 = mutTrace.2$GCA,
  GCC.1 = mutTrace.1$GCC,
  GCC.2 = mutTrace.2$GCC,
  GCG.1 = mutTrace.1$GCG,
  GCG.2 = mutTrace.2$GCG
)

posterior <- as.matrix(alanine)

mcmc_areas(posterior, prob = 0.8) + 
  ggtitle("Alanine Posterior Distributions","with medians and 80% intervals")


```

Now, to calculate some correlations between codons on different strands.

```{r}

#Testing a visualization method
ggplot() +
  geom_point(aes(x = mutTrace.1$GCA, y = mutTrace.2$GCA)) +
  geom_smooth(aes(x = mutTrace.1$GCA, y = mutTrace.2$GCA), method = "lm", color = "blue", se = TRUE) # Correlation line

regression <- lm(mutTrace.2$GCA ~ mutTrace.1$GCA) 
summary(regression)

#Calculate the correlation between each 'pair' of codons (mixture1:mixture2)
correlations <- sapply(names(mutTrace.1), function(col) {
  cor(mutTrace.1[[col]], mutTrace.2[[col]])
})

codon.corr <- data.frame(
  codon = names(correlations),
  corr = correlations,
  r.sq = correlations^2
)

hist(codon.corr$corr)
hist(codon.corr$r.sq)

print(codon.corr)

# Run regression analysis and store results
regression_results <- sapply(names(mutTrace.1), function(col) {
  model <- lm(mutTrace.2[[col]] ~ mutTrace.1[[col]])
  summary(model)$coefficients[2, ]  # Get the coefficient and its statistics
})

# Create a data frame from the results
regression_df <- data.frame(
  Codon = names(mutTrace.1),
  Estimate = regression_results[1, ],
  Std.Error = regression_results[2, ],
  t.value = regression_results[3, ],
  p.value = regression_results[4, ]
)

print(regression_df)

```

Let's examine our phi values

```{r}

#Now, back to our trace
phi_trace <- trace$getSynthesisRateTrace()[[1]]
names(phi_trace) <- getNames(genomes)
phi.trace <- phi_trace %>% data.frame() %>% slice(-1:-2000)
print(mean(colMeans(phi.trace, na.rm = TRUE)))
print(median(colMeans(phi.trace, na.rm = TRUE)))
print(range(colMeans(phi.trace, na.rm = TRUE)))

#Plotting phi distribution across genes
phi.trace.long <- phi.trace %>% pivot_longer(cols = 1:ncol(.), names_to = "gene", values_to = "phi") %>%
  mutate(phi.log = log(phi))
hist(phi.trace.long$phi)
hist(phi.trace.long$phi.log)

```

Now, let's view some empirical data. Importing RNA-seq data from:

Bhatia, Rama P., et al. “Transcriptomic Profiling of Escherichia Coli K-12 in Response to a Compendium of Stressors.” Scientific Reports, vol. 12, no. 1, May 2022, p. 8788. DOI.org (Crossref), <https://doi.org/10.1038/s41598-022-12463-3>.

```{r}

#Now, import ecoli expression data [Bhatia et al. 2022]
ecoli.data <- load_file("ecoli_expression_data.csv")
ecoli.data[ecoli.data == 0] <- NA
#ecoli.data[ecoli.data < 1] <- NA

e <- ecoli.data %>% select(-Locus_tag) %>% distinct() %>%
  mutate_at(.vars = 2:ncol(.), log) %>% column_to_rownames(var = "Gene_name")

#Plotting phi distribution across genes and calculating mean.normed data
e.long <- e %>% pivot_longer(cols = 1:ncol(.), names_to = "sample", values_to = "abundance") %>%
  mutate(m.norm = abundance / mean(abundance, na.rm = TRUE))

hist(e.long$abundance)
hist(e.long$m.norm)

print(mean(e.long$m.norm, na.rm = TRUE))
print(median(e.long$m.norm, na.rm = TRUE))
print(range(e.long$m.norm, na.rm = TRUE))

```

Other plots

```{r}

#Plot selection on y-axis and mutation on x-axis for each codon
selectionTrace <- trace$getCodonSpecificParameterTrace(1)
sel_Trace <- selectionTrace[[1]]
names(sel_Trace) <- AA.df$Codon

#Making trace data frames and discard burn-in
selTrace <- sel_Trace %>% data.frame() %>% slice(-1:-2000)

#Overlayed plots
ggplot() +
  # First scatterplot
  geom_point(aes(x = mutTrace.1$GCA, y = selTrace$GCA), color = "red", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.1$GCA, y = selTrace$GCA), method = "lm", color = "red", se = TRUE) +
  
  # Second scatterplot
  geom_point(aes(x = mutTrace.2$GCA, y = selTrace$GCA), color = "blue", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.2$GCA, y = selTrace$GCA), method = "lm", color = "blue", se = TRUE) +
  
  labs(title = "Selection and Mutation Trace: GCA", x = expression(Delta * M), y = expression(Delta * n),
       subtitle = "\tRed:leading | Blue:lagging")


#Overlayed plots
ggplot() +
  # First scatterplot
  geom_point(aes(x = mutTrace.1$TTC, y = selTrace$TTC), color = "red", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.1$TTC, y = selTrace$TTC), method = "lm", color = "red", se = TRUE) +
  
  # Second scatterplot
  geom_point(aes(x = mutTrace.2$TTC, y = selTrace$TTC), color = "blue", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.2$TTC, y = selTrace$TTC), method = "lm", color = "blue", se = TRUE) +
  
  labs(title = "Selection and Mutation Trace: TTC", x = expression(Delta * M), y = expression(Delta * n),
       subtitle = "\tRed:leading | Blue:lagging")


#Overlayed plots
ggplot() +
  # First scatterplot
  geom_point(aes(x = mutTrace.1$TTC, y = selTrace$TTC), color = "red", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.1$TTC, y = selTrace$TTC), method = "lm", color = "red", se = TRUE) +
  
  # Second scatterplot
  geom_point(aes(x = mutTrace.2$TTC, y = selTrace$TTC), color = "blue", alpha = 0.6) +
  geom_smooth(aes(x = mutTrace.2$TTC, y = selTrace$TTC), method = "lm", color = "blue", se = TRUE) +
  
  labs(title = "Selection and Mutation Trace: TTC", x = expression(Delta * M), y = expression(Delta * n),
       subtitle = "\tRed:leading | Blue:lagging")

```

